spacy/cli/find_threshold.py

```diff
@@ -0,0 +1,173 @@
+    beta (float): Beta for F-score calculation.
+    code_path (Optional[Path]): Path to Python file with additional code (registered functions) to be imported.
+    use_gpu (int): GPU ID or -1 for CPU.
+    gold_preproc (bool): Whether to use gold preprocessing. Gold preprocessing helps the annotations align to the
+        tokenization, and may result in sequences of more consistent length. However, it may reduce runtime accuracy due
+        to train/test skew.
+    silent (bool): Display more information for debugging purposes
+    """
+
+    util.logger.setLevel(logging.DEBUG if verbose else logging.INFO)
+    import_code(code_path)
+    find_threshold(
+        model=model,
+        data_path=data_path,
+        pipe_name=pipe_name,
+        threshold_key=threshold_key,
+        scores_key=scores_key,
+        n_trials=n_trials,
+        beta=beta,
+        use_gpu=use_gpu,
+        gold_preproc=gold_preproc,
+        silent=False,
+    )
+
+
+def find_threshold(
+    model: str,
+    data_path: Path,
+    pipe_name: str,
+    threshold_key: str,
+    scores_key: str,
+    *,
+    n_trials: int = _DEFAULTS["n_trials"],  # type: ignore
+    beta: float = _DEFAULTS["beta"],  # type: ignore
+    use_gpu: int = _DEFAULTS["use_gpu"],  # type: ignore
+    gold_preproc: bool = _DEFAULTS["gold_preproc"],  # type: ignore
+    silent: bool = True,
+) -> Tuple[float, float]:
+    """
+    Runs prediction trials for `textcat` models with varying tresholds to maximize the specified metric.
+    model (Union[str, Path]): Path to file with trained model.
+    data_path (Union[str, Path]): Path to file with DocBin with docs to use for threshold search.
+    pipe_name (str): Name of pipe to examine thresholds for.
+    threshold_key (str): Key of threshold attribute in component's configuration.
+    scores_key (str): Name of score to metric to optimize.
+    n_trials (int): Number of trials to determine optimal thresholds.
+    beta (float): Beta for F-score calculation.
+    use_gpu (int): GPU ID or -1 for CPU.
+    gold_preproc (bool): Whether to use gold preprocessing. Gold preprocessing helps the annotations align to the
+        tokenization, and may result in sequences of more consistent length. However, it may reduce runtime accuracy due
+        to train/test skew.
+    silent (bool): Whether to print non-error-related output to stdout.
+    RETURNS (Tuple[float, float]): Best found threshold with corresponding F-score.
+    """
+
+    setup_gpu(use_gpu, silent=silent)
+    data_path = util.ensure_path(data_path)
+    if not data_path.exists():
+        wasabi.msg.fail("Evaluation data not found", data_path, exits=1)
+    nlp = util.load_model(model)
+
+    pipe: Optional[Pipe] = None
+    try:
+        pipe = nlp.get_pipe(pipe_name)
+    except KeyError as err:
+        wasabi.msg.fail(title=str(err), exits=1)
+
+    if not isinstance(pipe, TrainablePipe):
+        raise TypeError(Errors.E1044)
```

Comment:
I don't see any reason for this restriction?

Reply:
`Pipe` doesn't necessarily have a `.cfg` attribute, which I assume to be available later on. I thought `mypy` complained about this otherwise, but apparently this isn't the case. Should I drop this?

Reply:
In an ideal world, settings would only be changed in `nlp.config` to modify the thresholds.

In the current actual v3 world, I think you would have to try to set this in both `nlp.config` and `pipe.cfg`.

Reply:
Setting `threshold` via `pipe.cfg` seems to work. What's the reason we'd do this in `nlp.config` as well?

Reply:
It's not that it doesn't work, it's that in an ideal world these settings would all be only in `nlp.config`. It's a carry-over from v2 that they're stored in `pipe.cfg` and it was a workaround in the scorer refactoring in v3.1 that the `pipe.cfg` settings are passed internally to the scorer.

Actually, let me think a second, you might have to reload the model for every pass to get this work with `nlp.config`, but this is much cleaner and easier to specify for the user. But it is currently bad that the settings are duplicated in `nlp.config` and `pipe.cfg`, and then saved and reloaded from `pipe.cfg` instead of `nlp.config`, since they can get out-of-sync.

Reply:
This is a better design (`pipe.cfg` is empty since everything is just in the config):

https://github.com/explosion/spacy-experimental/blob/7ca435d01aa87b2e61760a266fe8cdb748043ee1/spacy_experimental/span_finder/span_finder_component.py#L45-L144

Reply:
What's the recommended way to update `nlp.config` programmatically? Took me a while to realize that `config` is rebuild in the getter. And `components` is rebuilt completely, so I'd have to change the content of `_pipe_configs`. I don't see how to though, other than deleting and adding the pipe.

Reply:
I think it will have to involve reloading the pipeline for each iteration, and you'd modify it with overrides in `spacy.load("model", config={...overrides...})`.

Reply:
I think this would be much easier if left until v4. If you do it now you'd have to do both the config and hack things into `pipe.cfg`.

Reply:
Pipeline is reloaded with overrides for the threshold on every threshold change now.

> I don't see any reason for this restriction?
Coming back to this: this is for proper typing and type hinting. `nlp.get_pipe()` has a signature specifying the return of a `Pipe` object, yet we access its `.cfg` attribute to override the threshold setting. And I stand corrected, `mypy` actually issues an error about this if this type check is removed.

Reply:
Then this should be refactored to work for both components without `.cfg` and non-`Pipe` components.

I think part of this is that the return type for `get_pipe` is incorrect. We don't actually require any particular type for pipeline components. I could also imagine non-trainable pipe components or other kinds of custom components that have thresholds.

Reply:
What would be our approach be to changing the threshold value for components without `.cfg`? Try to manipulate `.threshold` directly and if that fails, abort?

Reply:
Modifying it in `.cfg` is basically a bad hack. The case that should always work is modifying it in the config.

Reply:
Not assuming `.cfg` or the component being a `Pipe` in the latest commit.